cpp/include/cudf/null_mask.hpp

```diff
@@ -24,6 +24,15 @@
 
 namespace cudf {
 
+/**
+ * @brief Bitmask output type.
+ */
+struct bitmask {
```

Comment:
Meh, seeing this, I agree with @bdice's earlier concern about the naming of this type. `bitmask` is a very loaded term through libcudf and this is an odd place to just drop a type named `bitmask`. 

I'd be tempted to just return a `tuple<device_buffer, int, int>`, but then you lose the ability to give the members meaningful names... I wish we could use an anonymous type here. That said, structured bindings makes it so we aren't really concerned with the names of the members...

Reply:
Or just return a `pair<buffer,int>` with the valid(null) count and the null(valid) count can just be computed.

Reply:
> Or just return a pair<buffer,int> with the valid(null) count and the null(valid) count can just be computed.

A pair would be my preference as well. I think the caller should always have access to the total count, so the other value is just a subtraction away.

Reply:
> the other value is just a subtraction

That's actually my concern, in case bitmask is `nullptr`, both `null_count`  and `valid_count` equal zero so the subtraction outside of `bitmask_and/or` without checking if bitmask is `nullptr` will not work and it's easy to misuse it.

Reply:
I thought this would work fine if we returned `std::pair<buffer, size_type>` where the second item is the _null count_ (not the valid count, which would cause issues with the `nullptr` case). I am not deeply familiar with the expected semantics but I would expect that if the bitmask is a `nullptr`, then `valid_count == size` (valid count is not equal to zero). If that's not true, why not? Shouldn't it be an invariant that `size == null_count + valid_count`?

Reply:
`size == null_count + valid_count` (which is an invariant) only if the bitmask is not a `nullptr`. If the mask is `nullptr`, `valid_count == null_count == 0`.

Reply:
Hmm. I agree that [`count_set_bits(nullptr)` returns zero](https://github.com/rapidsai/cudf/blob/27184431443fad3207590a10d9e32144d7c4c827/cpp/src/bitmask/null_mask.cu#L336-L341), which makes sense as a standalone decision for generic bitmasks. However, in the context of bitmasks _representing item validity_, this gets murkier.

I don't see a "valid count" method on a column that mirrors the [`null_count` method](https://github.com/rapidsai/cudf/blob/b47ff2408a662e0ef3885187a83ff4ad2db19d6a/cpp/src/column/column.cu#L139-L148), which would clearly define this invariant for columns with masks that are `nullptr`. Basically, I am not sure if we have consistent semantics around `nullptr` implying `valid_count == 0` -- only the implementation of `count_set_bits` leads me in that direction. Here are some counterexamples:

Python seems to have the invariant `size == valid_count + null_count` (even for non-nullable columns as far as I can tell): https://github.com/rapidsai/cudf/blob/b47ff2408a662e0ef3885187a83ff4ad2db19d6a/python/cudf/cudf/core/column/column.py#L396-L399

Here are cases where the valid (or null) count is computed by subtraction regardless of `col.has_nulls()`:
https://github.com/rapidsai/cudf/blob/27184431443fad3207590a10d9e32144d7c4c827/cpp/src/reductions/compound.cuh#L52
https://github.com/rapidsai/cudf/blob/ce385ce4244c19d37debdb51d9f5fe726d8665f3/cpp/src/rolling/rolling_detail.cuh#L1055

This makes it seem like we rely (perhaps inconsistently?) on this invariant `size == null_count + valid_count` holding even if the mask is `nullptr`.

Can you help me find examples or documentation supporting the claim that this invariant is only meaningful when `mask != nullptr`?

Reply:
I think I know what should be the proper way to do this then. Currently, there is no use case of valid count and the first one would be in an upcoming hash join optimization PR. Overall, my previous statement about invariant was wrong and it should be meaningful no matter how the mask state changes. `bitmask_and/or` should return a pair of resulting mask and null count.

Reply:
Awesome! I agree with that plan as the clearest path forward. Thanks for talking through that together — I appreciate your insight and thoughtful responses! 👍

Reply:
Voilà! Thanks for the thorough review! I learned a lot.